71_RayTracingPipeline/app_resources/common.hlsl

```diff
@@ -4,42 +4,109 @@
 #include "nbl/builtin/hlsl/cpp_compat.hlsl"
 
 NBL_CONSTEXPR uint32_t WorkgroupSize = 16;
+NBL_CONSTEXPR uint32_t MAX_UNORM_10 = 1023;
+NBL_CONSTEXPR uint32_t MAX_UNORM_22 = 4194303;
+
+inline uint32_t packUnorm10(float32_t v)
+{
+    return trunc(v * float32_t(MAX_UNORM_10) + 0.5f);
+}
+
+inline float32_t unpackUnorm10(uint32_t packed)
+{
+    return float32_t(packed & 0x3ff) * (1.0f / float32_t(MAX_UNORM_10));
+}
+
+inline uint32_t packUnorm22(float32_t v)
+{
+    const float maxValue = float32_t(MAX_UNORM_22);
+    return trunc(v * maxValue + 0.5f);
+}
+
+inline float32_t unpackUnorm22(uint32_t packed)
+{
+    const float maxValue = float32_t(MAX_UNORM_22);
+    return float32_t(packed & 0x3fffff) * (1.0f / maxValue);
+}
+
+inline uint32_t packUnorm3x10(float32_t3 v)
+{
+    return (packUnorm10(v.z) << 20 | (packUnorm10(v.y) << 10 | packUnorm10(v.x)));
+}
+
+inline float32_t3 unpackUnorm3x10(uint32_t packed)
+{
+    return float32_t3(unpackUnorm10(packed), unpackUnorm10(packed >> 10), unpackUnorm10(packed >> 20));
+}
 
 struct Material
 {
 	float32_t3 ambient;
     float32_t3 diffuse;
     float32_t3 specular;
     float32_t shininess;
-    float32_t dissolve; // 1 == opaque; 0 == fully transparent
-    uint32_t illum; // illumination model (see http://www.fileformat.info/format/material/)
+    float32_t alpha;
+
+    bool isTransparent() NBL_CONST_MEMBER_FUNC
+    {
+        return alpha < 1.0;
+    }
 };
 
-struct SProceduralGeomInfo
+struct MaterialPacked
 {
-    float32_t3 center;
-    float32_t radius;
-    Material material;
+	uint32_t ambient;
+    uint32_t diffuse;
+    uint32_t specular;
+    uint32_t shininess: 22;
+    uint32_t alpha : 10;
+
+    bool isTransparent() NBL_CONST_MEMBER_FUNC
+    {
+        return alpha != MAX_UNORM_10;
+}
 };
 
-struct Aabb
+inline MaterialPacked packMaterial(Material material)
+{
+    MaterialPacked packed;
+    packed.ambient = packUnorm3x10(material.ambient);      
+    packed.diffuse = packUnorm3x10(material.diffuse);
+    packed.specular = packUnorm3x10(material.specular);      
+    packed.shininess = packUnorm22(material.shininess);
+    packed.alpha = packUnorm10(material.alpha);
+    return packed;
+}
+
+inline Material unpackMaterial(MaterialPacked packed)
```

Comment:
style nitpick, a specialization of `nbl:hlsl::impl::static_cast_helper` would have been nicer